### ISSUE TITLE
Add metadata capabilities to urql

### ISSUE THREAD

**Issue by andyrichardson** [FormidableLabs/urql#291]
In order to expose enough information to create valuable devtools I'm proposing we add a _meta_ tag to our operations.

This tag can be set/updated at the following points:
 - on calling hooks / client execution
 - inside hooks (for sourcing the calling component)
 - inside any exchange

We could use context but I think keeping it separate (such as a property within context as @kitten has suggested) would be best as any information added to the meta property of an operation should not be mission critical (i.e. no worry of breaking anything at runtime).

**tl;dr:** add meta property to operations

Edit: clear up wording on using context

**Comment by kitten** [comment]
Do you want to summarise what you need. I'd be careful to add it to every exchange as that will make it a lot harder to write exchanges.

But what info do we really need? Whether some result has been retrieved from cache?

It'd be a bit easier to think about how it'd look if we know what info we need 😅

**Comment by andyrichardson** [comment]
@kitten the point with this proposal is to add an ability to add _any_ metadata to an operation at any point in an operations lifecycle.

While there are specific things we need this for (i.e. devtools) this is more intended to be a intent-agnostic metadata layer. As people create their own exchanges, the uses for this will vary.

**Comment by kitten** [comment]
@andyrichardson the thing though is that we have the `context` to add any data that exchanges may need. 😅

I see this information as targeted specifically at the devtools. Now we could add `context.meta` for instance, but it'd be interesting to discuss what information the devtools need

**Comment by andyrichardson** [comment]
@kitten I already acknowledged the availability of context in the issue. 

The incentives behind not just using context are:
 - metadata additions should not be able to break functionality
 - we need a separation between what is causing effects and what is purely metadata
 - we need to expose metadata to the end user for hooks/execution functions. Adding the ability to override the whole context would be a terrible idea.

> Now we could add `context.meta`

Yes, storing the metadata inside the context works fine - using context for metadata would not be fine. The proposal is to pass metadata throughout an operations lifecycle.

The information we need for devtools isn't what I'm proposing here, it's the ability to pass metadata through an operation's lifecycle. I'm getting the impression you're not a fan. What's the concern with this addition?

**Comment by kitten** [comment]
@andyrichardson I think `context.meta` would be the best place for this addition 👍

I can't say whether I'm a fan or not because I'm unsure what information we'd actually add there 😅 the lifecycle is currently quite controlled via the input / output. So in the client we already know which queries are ongoing / subscribed to, what operations have started or got results, and whether that came from a cache or not (or rather resolved synchronously or not)

So I'm just a bit confused. Totally fine with reserving a meta field on context or the operation for this. But I just don't know what we'd use it for yet 😅

**Comment by andyrichardson** [comment]
> I think context.meta would be the best place for this addition 👍

Agreed! I'll put something together over the coming days 🚀 

> I can't say whether I'm a fan or not because I'm unsure what information we'd actually add there 

I hear you! While this will be super valuable for devtools, the last thing we want to do is write a devtools solution to a problem that others may want to solve themselves. 

> So in the client we already know which queries are ongoing / subscribed to, what operations have started or got results, and whether that came from a cache or not (or rather resolved synchronously or not)

💯 but there's tons of information we are yet to get which could be useful such as:
 - The source component for a given operation
 - The state of the cache after a given operation
 - Who knows what else!

**Comment by kitten** [comment]
Well, I do see that we'll need it eventually 😅 but you can get the source component by calling the React utility for it whenever the operation comes in as long as that's synchronous.

I do totally see value in more information from the cache. But we probably need to be more granular than per operation.

Anyway, let's see what we need to add 🙌
